### ISSUE TITLE
compression middleware

### ISSUE THREAD

**Issue by seiyria** [SocketCluster/socketcluster#162]
Is it feasible or sensible to use middleware to run some form of data compression between client and server? What would be the recommended way to compress payloads between client and server? Is there any compression currently?

**Comment by jondubois** [comment]
@seiyria Only if your bottleneck is bandwidth. Compression/decompression uses extra computational resources. It might be useful maybe if you need to deliver a small number of very popular resources in realtime - Then you could store the compressed versions of these resources in cache for example.

This could make sense if I had a scenario involving very low throughput channels each publishing very large JSON or binary objects to millions of subscribers.

**Comment by seiyria** [comment]
Well, the situation I'm in is that I'm sending a lot of data to a client. Clients are on mobile devices and my server is hosted on heroku. Some of the things I send via socket are large sets of data, and I'm trying to lower the size that I send. I'm not worried about computational power on client or server right now because the target devices are pretty powerful and the server is scalable (usually network connectivity is the problem). Because of this, I figured that it would be wise to send either jwt-compressed bits or something.

To be specific on what exactly I'm making: I have monsters that are generated for players. They can be a bit large, data-wise (though I'll work on compressing them). I send 650-850 of them to a client per request (generally, per hour or per login). I figured generating a jwt token per monster would be better for sending them in that quantity, instead of sending the raw object. They look like this:

```
{
            "name": "Goblin",
            "profession": "Monster",
            "professionLevels": {
                "Monster": 1,
                "Thief": 1,
                "Fighter": 1,
                "Mage": 1
            },
            "statusEffects": [],
            "stats": {
                "gold": 0,
                "xp": {},
                "hp": {
                    "minimum": 0,
                    "maximum": 5,
                    "__current": 5,
                    "booster": 0
                },
                "mp": {
                    "minimum": 0,
                    "maximum": 3,
                    "__current": 3,
                    "booster": 0
                },
                "str": 1,
                "vit": 9,
                "dex": -1,
                "mnt": 2,
                "luk": 0,
                "acc": -1
            },
            "skills": [],
            "inventory": [],
            "equipment": {
                "weapon": {
                    "itemId": "41eba92d-7aab-4cd9-aa1b-7bcdd83f307e",
                    "name": "Claw",
                    "stats": {},
                    "levelRequirement": 1,
                    "quality": 0,
                    "value": 0,
                    "type": "weapon"
                },
                "armor": {
                    "itemId": "9a268717-142d-4333-a75c-379fde317ddb",
                    "name": "Fullplate",
                    "stats": {
                        "vit": 8,
                        "dex": -2,
                        "acc": -1
                    },
                    "levelRequirement": 1,
                    "quality": 0,
                    "value": 23,
                    "type": "armor",
                    "dropRate": 25
                },
                "profession": {
                    "stats": {
                        "str": 1,
                        "mnt": 2,
                        "dex": 1,
                        "vit": 1,
                        "luk": 0.3,
                        "acc": 0
                    }
                },
                "buffs": {
                    "stats": {}
                }
            },
            "unlockedProfessions": [
                "Thief",
                "Fighter",
                "Mage"
            ],
            "cooldowns": {},
            "slug": "MON-1",
            "id": "58bebdd0-ab05-43a2-9b05-7ed0e3db7646",
            "bonusXp": 0,
            "seed": 1457067600563,
            "verifyToken": "049c7cb45bbc5a1025195e3a66f62974",
            "location": {
                "lat": 44.031542946243775,
                "lon": -88.55779830910286
            },
            "rating": -2,
            "goldDrop": "1d4"
        }
```

In this, I generate a `verifyToken` based on some data so the client can't spoof any data about a monster. They send the monster + the token back to the server and the server verifies that the monster was generated correctly. I might be doing it entirely incorrectly, though.

Unfortunately none of this data is cached, it's recalculated hourly also. Each client has  their own unique set of data that's calculated hourly, on top of that.

**Comment by seiyria** [comment]
Alright. I squared away my issue. Surprisingly it was a rendering issue in testing, though I haven't determined if bandwidth or rendering will be the problem in production. If it is, I found that JSONPack gives me approximately 40% savings in compression of 750 of the above objects, so that's the route I'll go. Not sure I'll use middleware since I don't think those can be added on the client(?) and it shouldn't be added for endpoints that aren't a lot of data, anyway.
